gcovr/utils.py

```diff
@@ -290,44 +293,44 @@ def match(self, path):
         return super(DirectoryPrefixFilter, self).match(normpath)
 
 
-class Logger(object):
-    def __init__(self, verbose=False):
-        self.verbose = verbose
+class MaximumApplicableLogLevelFilter(Filter):
+    """Log messages of the given level and lower are passed to the handler"""
+    def __init__(self, level):
+        self.level = level
+
+    def filter(self, record):
+        return record.levelno <= self.level
 
-    def warn(self, pattern, *args, **kwargs):
-        """Write a formatted warning to STDERR.
 
-        pattern: a str.format pattern
-        args, kwargs: str.format arguments
-        """
-        pattern = "(WARNING) " + pattern + "\n"
-        sys.stderr.write(pattern.format(*args, **kwargs))
+def configure_logging(logger: logging.Logger, level: int = logging.INFO) -> None:
+    # stdout configuration
+    stdout_handler = logging.StreamHandler(sys.stdout)
+    info_and_lower = MaximumApplicableLogLevelFilter(logging.INFO)
+    stdout_handler.addFilter(info_and_lower)
+    stdout_handler.setFormatter(logging.Formatter("%(message)s"))
+    stdout_handler.setLevel(level)
 
-    def error(self, pattern, *args, **kwargs):
-        """Write a formatted error to STDERR.
+    # stderr warning configuration
+    stderr_warning_handler = logging.StreamHandler(sys.stderr)
+    warning_and_lower = MaximumApplicableLogLevelFilter(logging.WARNING)
+    stdout_handler.addFilter(warning_and_lower)
+    stderr_warning_handler.setFormatter(logging.Formatter("(WARNING) %(message)s"))
+    stderr_warning_handler.setLevel(logging.WARNING)
 
-        pattern: a str.format pattern
-        args, kwargs: str.format parameters
-        """
-        pattern = "(ERROR) " + pattern + "\n"
-        sys.stderr.write(pattern.format(*args, **kwargs))
+    # stderr warning configuration
+    stderr_error_handler = logging.StreamHandler(sys.stderr)
+    stderr_error_handler.setFormatter(logging.Formatter("(ERROR) %(message)s"))
+    stderr_error_handler.setLevel(logging.ERROR)
 
-    def msg(self, pattern, *args, **kwargs):
-        """Write a formatted message to STDOUT.
+    logger.addHandler(stdout_handler)
+    logger.addHandler(stderr_warning_handler)
+    logger.addHandler(stderr_error_handler)
 
-        pattern: a str.format pattern
-        args, kwargs: str.format arguments
-        """
-        pattern = pattern + "\n"
-        sys.stdout.write(pattern.format(*args, **kwargs))
+    sys.excepthook = exception_hook
 
-    def verbose_msg(self, pattern, *args, **kwargs):
-        """Write a formatted message to STDOUT if in verbose mode.
 
-        see: self.msg()
-        """
-        if self.verbose:
-            self.msg(pattern, *args, **kwargs)
+def exception_hook(exc_type, exc_value, exc_traceback) -> None:
+    logging.exception("Uncaught EXCEPTION", exc_info=(exc_type, exc_value, exc_traceback))
```

Comment:
Is this complicated configuration really needed? Logging is always done to `stderr` and the message can be configured with the format:
```python
"(%(levelname)s) %(message)s"
```

Reply:
This logging configuration replicates the old Logger class functionality completely. Look at the actual implementation. So splitting between `stderr` and `stdout` is required, as well that different log levels use a different message format. Unless you want to change the tests as well, this has to be done like this. At least as first step, to avoid too many changes at ones.

Otherwise, `stderr` and  `"(%(levelname)s) %(message)s"` would simplify things a bit.

Reply:
> It could make sense to always print these to stdout directly, without involving the logger.

Configur logger in main function to always print to `stderr`. The only change here is that infos get a `INFO` prepended.

Reply:
Changed this to always log to `stderr` and using `"(%(levelname)s) %(message)s"` to always add the level in the output.